README.md
# DDE Control Center

The control center of Deepin Desktop Environment 2.0 (DDE)

## Dependencies

* Qt5: 5.3.1 or later and the following modules are required
    * Core
    * DBus
    * Gui
    * Qml
    * Quick
    * Widgets
    * Svg
    * Webkit
    * Multimedia

* Others:
    * dde-account-faces
    * dde-daemon
    * dde-api
    * dde-qml-dbus-factory
    * dde-qml-dbus-factory
    * deepin-artwork-themes
    * deepin-qml-widgets
    * libgtk2.0
    * libx11


## Installation

```
mkdir build; cd build
qmake ..
make && make install
```

## Usage

`dde-control-center show`

## Getting help

Any usage issues can ask for help via

* [Gitter](https://gitter.im/orgs/linuxdeepin/rooms)
* [IRC channel](https://webchat.freenode.net/?channels=deepin)
* [Forum](https://bbs.deepin.org)
* [WiKi](http://wiki.deepin.org/)

## Getting involved

We encourage you to report issues and contribute changes

* [Contribution guide for users](http://wiki.deepin.org/index.php?title=Contribution_Guidelines_for_Users)
* [Contribution guide for developers](http://wiki.deepin.org/index.php?title=Contribution_Guidelines_for_Developers).

## License

DDE Control Center is licensed under [GPLv3](LICENSE).


debian/changelog
dde-control-center (5.6.0102.1+icbc) unstable; urgency=low

  * Autobuild Tag 3.0.14 

 -- TagBuilder <tagbuilder@deepin.com>  Fri, 12 Aug 2016 18:43:22 +0000

dde-control-center (3.0.13-1) unstable; urgency=low

  * Autobuild Tag 3.0.13 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Thu, 26 May 2016 17:24:07 +0000

dde-control-center (3.0.12-1) unstable; urgency=low

  * Autobuild Tag 3.0.12 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Mon, 23 May 2016 14:49:33 +0000

dde-control-center (3.0.10-7) unstable; urgency=low

  * Rebuild Version 3.0.10 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Mon, 25 Apr 2016 11:20:16 +0000

dde-control-center (3.0.10-6) unstable; urgency=low

  * Rebuild Version 3.0.10 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Thu, 21 Apr 2016 16:16:22 +0000

dde-control-center (3.0.10-5) unstable; urgency=low

  * Rebuild Version 3.0.10 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Wed, 09 Mar 2016 15:25:09 +0000

dde-control-center (3.0.10-4) unstable; urgency=low

  * Rebuild Version 3.0.10 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Wed, 09 Mar 2016 12:33:35 +0000

dde-control-center (3.0.10-3) unstable; urgency=low

  * Rebuild Version 3.0.10 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Wed, 09 Mar 2016 11:38:29 +0000

dde-control-center (3.0.10-2) unstable; urgency=low

  * Rebuild Version 3.0.10 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Wed, 09 Mar 2016 11:07:31 +0000

dde-control-center (3.0.10-1) unstable; urgency=low

  * Autobuild Tag 3.0.10 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Mon, 07 Mar 2016 17:01:12 +0000

dde-control-center (3.0.9-3) stable; urgency=low

  * Rebuild Version 3.0.9 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Thu, 25 Feb 2016 16:16:18 +0000

dde-control-center (3.0.9-2) stable; urgency=low

  * Rebuild Version 3.0.9 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Thu, 25 Feb 2016 11:24:31 +0000

dde-control-center (3.0.9-1) stable; urgency=low

  * Autobuild Tag 3.0.9 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Thu, 25 Feb 2016 11:09:23 +0000

dde-control-center (3.0.8-1) stable; urgency=low

  * Autobuild Tag 3.0.8 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Mon, 22 Feb 2016 17:01:22 +0000

dde-control-center (3.0.7-1) stable; urgency=low

  * Autobuild Tag 3.0.7 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Wed, 27 Jan 2016 18:51:35 +0000

dde-control-center (3.0.6-1) stable; urgency=low

  * Autobuild Tag 3.0.6 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Fri, 22 Jan 2016 20:10:14 +0000

dde-control-center (3.0.5-3) stable; urgency=low

  * Rebuild Version 3.0.5 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Fri, 15 Jan 2016 17:44:02 +0000

dde-control-center (3.0.5-2) stable; urgency=low

  * Rebuild Version 3.0.5 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Fri, 15 Jan 2016 16:02:54 +0000

dde-control-center (3.0.5-1) stable; urgency=low

  * Autobuild Tag 3.0.5 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Fri, 15 Jan 2016 15:44:01 +0000

dde-control-center (3.0.4-1) stable; urgency=low

  * Autobuild Tag 3.0.4 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Tue, 12 Jan 2016 20:08:17 +0000

dde-control-center (3.0.3-1) stable; urgency=low

  * Autobuild Tag 3.0.3 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Wed, 06 Jan 2016 16:42:35 +0000

dde-control-center (3.0.2-1) stable; urgency=low

  * Autobuild Tag 3.0.2 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Wed, 30 Dec 2015 15:16:56 +0000

dde-control-center (3.0.1-1) stable; urgency=low

  * Autobuild Tag 3.0.1 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Tue, 29 Dec 2015 21:43:24 +0000

dde-control-center (3.0.0-1) stable; urgency=low

  * Autobuild Tag 3.0.0 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Mon, 28 Dec 2015 21:15:54 +0000

dde-control-center (2.91.6-2) stable; urgency=low

  * Rebuild Version 2.91.6 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Tue, 22 Dec 2015 13:22:00 +0000

dde-control-center (2.91.6-1) stable; urgency=low

  * Autobuild Tag 2.91.6 

 -- TagBuilder <tagbuilder@linuxdeepin.com>  Mon, 21 Dec 2015 22:37:34 +0000

dde-control-center (2.91.5-2) stable; urgency=low

  * Rebuild Version 2.91.5 

 -- TagBuilder <tabbuilder@linuxdeepin.com>  Sun, 20 Dec 2015 12:43:15 +0000

dde-control-center (2.91.5-1) stable; urgency=low

  * Autobuild Tag 2.91.5 

 -- TagBuilder <tabbuilder@linuxdeepin.com>  Sat, 19 Dec 2015 15:58:24 +0000

dde-control-center (2.91.4-1) stable; urgency=low

  * Autobuild Tag 2.91.4 

 -- TagBuilder <tabbuilder@linuxdeepin.com>  Thu, 10 Dec 2015 20:33:06 +0000

dde-control-center (2.91.3-1) stable; urgency=low

  * Autobuild Tag 2.91.3 

 -- TagBuilder <tabbuilder@linuxdeepin.com>  Wed, 09 Dec 2015 22:25:45 +0000

dde-control-center (2.91.2-1) stable; urgency=low

  * Autobuild Tag 2.91.2 

 -- TagBuilder <tabbuilder@linuxdeepin.com>  Fri, 20 Nov 2015 17:02:56 +0000

dde-control-center (2.91.1-1) stable; urgency=low

  * Autobuild Tag 2.91.1 

 -- TagBuilder <tabbuilder@linuxdeepin.com>  Fri, 20 Nov 2015 09:52:29 +0000

dde-control-center (2.90.1+r1494~g999cab5) experimental; urgency=low

  ** Build revision: 999cab57c67c88f9a9065963c61c0cd262897918  @  origin/release/2.3 **

  [upstreamer]
  * Use DPalette to replace DConstants
  * Hide remoting module for 2015 test
  * Show error message when failed to connect to remote peer
  * Fixed: show error page on peerIdFailed
  * Merge branch 'develop' into release/2.3

 -- None <None>  Thu, 02 Jul 2015 13:55:11 +0000

dde-control-center (0.0.0) unstable; urgency=medium

  * Unstable release.

 -- 宋文武 <songwenwu@linuxdeepin.com>  Tue, 02 Dec 2014 10:19:15 +0800


frame/main.cpp
/**
 * Copyright (C) 2015 Deepin Technology Co., Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 **/

#include "frame.h"
#include "dbuscontrolcenterservice.h"

#include <DApplication>
#include <DLog>
#include <QStyle>

DWIDGET_USE_NAMESPACE
#ifdef DCORE_NAMESPACE
DCORE_USE_NAMESPACE
#else
DUTIL_USE_NAMESPACE
#endif

static const QString getQssFromFile(const QString &name)
{
#ifdef QT_DEBUG
//    qDebug() << "load qss file: " << name;
#endif

    QString qss;

    QFile f(name);
    if (f.open(QIODevice::ReadOnly | QIODevice::Text)) {
        qss = f.readAll();
        f.close();
    }

    return qss;
}

static const QString getStyleSheetFromDir(QDir dir)
{
    QString ret;

    for (auto name : dir.entryInfoList(QDir::Files | QDir::NoDotAndDotDot)) {
        if (name.suffix() == "qss") {
            ret.append(getQssFromFile(name.absoluteFilePath()));
        }
    }

    return ret;
}

static const QString styleSheetFromTheme(const QString &theme)
{
    QStringList moduleList = { "frame", "widgets", "accounts", "defapp",
                               "mouse", "wacom", "sound", "bluetooth",
                               "personalization", "datetime", "display",
                               "keyboard", "power", "systeminfo",  "update", "network"};

    QString ret;

    const QString resources = ":/%1/themes/" + theme;
    for (auto module : moduleList)
    {
        QString dir = resources.arg(module);
        ret.append(getStyleSheetFromDir(dir));
    }

    return ret;
}

static void onThemeChange(const QString &theme)
{
    QString qss;
    qss.append(styleSheetFromTheme("common"));
    qss.append(styleSheetFromTheme(theme));

#ifdef QT_DEBUG
//    qDebug() << "set theme:" << theme;
//    qDebug().noquote() << "qss: " << endl << qss;
#endif

    qApp->setStyleSheet(qss);
}

static void onFontSizeChanged(const float pointSizeF) {
    // TODO(hualet): DPI default to 96.
    auto PxToPt = [] (int px) -> float {
        return px * 72.0 / 96;
    };

    QFont font(qApp->font());
    if (pointSizeF <= 8.2) {
        font.setPointSizeF(PxToPt(10));
        qApp->setFont(font, "dcc::widgets::SmallLabel");
        font.setPointSizeF(PxToPt(11));
        qApp->setFont(font, "dcc::widgets::NormalLabel");
        font.setPointSizeF(PxToPt(12));
        qApp->setFont(font, "dcc::widgets::LargeLabel");
    } else if (pointSizeF == 9) {
        font.setPointSizeF(PxToPt(11));
        qApp->setFont(font, "dcc::widgets::SmallLabel");
        font.setPointSizeF(PxToPt(12));
        qApp->setFont(font, "dcc::widgets::NormalLabel");
        font.setPointSizeF(PxToPt(16));
        qApp->setFont(font, "dcc::widgets::LargeLabel");
    }  else if (pointSizeF == 9.7) {
        font.setPointSizeF(PxToPt(11));
        qApp->setFont(font, "dcc::widgets::SmallLabel");
        font.setPointSizeF(PxToPt(13));
        qApp->setFont(font, "dcc::widgets::NormalLabel");
        font.setPointSizeF(PxToPt(15));
        qApp->setFont(font, "dcc::widgets::LargeLabel");
    } else if (pointSizeF == 11.2) {
        font.setPointSizeF(PxToPt(12));
        qApp->setFont(font, "dcc::widgets::SmallLabel");
        font.setPointSizeF(PxToPt(15));
        qApp->setFont(font, "dcc::widgets::NormalLabel");
        font.setPointSizeF(PxToPt(18));
        qApp->setFont(font, "dcc::widgets::LargeLabel");
    } else if (pointSizeF >= 12) {
        font.setPointSizeF(PxToPt(14));
        qApp->setFont(font, "dcc::widgets::SmallLabel");
        font.setPointSizeF(PxToPt(16));
        qApp->setFont(font, "dcc::widgets::NormalLabel");
        font.setPointSizeF(PxToPt(20));
        qApp->setFont(font, "dcc::widgets::LargeLabel");
    }
}

int main(int argc, char *argv[])
{
    DApplication::loadDXcbPlugin();

    DApplication app(argc, argv);
    app.setOrganizationName("deepin");
    app.setApplicationName("dde-control-center");
    app.setApplicationVersion("4.0");
    app.loadTranslator();

    DLogManager::registerConsoleAppender();
    DLogManager::registerFileAppender();

    qApp->setQuitOnLastWindowClosed(false);

    // take care of command line options
    QCommandLineOption showOption(QStringList() << "s" << "show", "show control center(hide for default).");
    QCommandLineOption toggleOption(QStringList() << "t" << "toggle", "toggle control center visible.");
    QCommandLineParser parser;
    parser.setApplicationDescription("DDE Control Center");
    parser.addHelpOption();
    parser.addVersionOption();
    parser.addOption(showOption);
    parser.addOption(toggleOption);
    parser.addPositionalArgument("module", "the module's id of which to be shown.");
    parser.process(app);

    app.setTheme("semidark");
    onThemeChange("dark");

    QTimer::singleShot(0, [] { onFontSizeChanged(qApp->font().pointSizeF()); });

    QStringList positionalArgs = parser.positionalArguments();

    Frame f;
    DBusControlCenterService adaptor(&f);
    Q_UNUSED(adaptor);
    QDBusConnection conn = QDBusConnection::sessionBus();
    if (!conn.registerService("com.deepin.dde.ControlCenter") ||
        !conn.registerObject("/com/deepin/dde/ControlCenter", &f)) {
        qDebug() << "dbus service already registered!";

        if (parser.isSet(toggleOption))
            QProcess::startDetached("dbus-send --print-reply --dest=com.deepin.dde.ControlCenter /com/deepin/dde/ControlCenter com.deepin.dde.ControlCenter.Toggle");

        if (!positionalArgs.isEmpty())
            QProcess::startDetached(QString("dbus-send --print-reply --dest=com.deepin.dde.ControlCenter /com/deepin/dde/ControlCenter com.deepin.dde.ControlCenter.ShowModule \"string:%1\"").arg(positionalArgs.first()));
        else if (parser.isSet(showOption))
            QProcess::startDetached("dbus-send --print-reply --dest=com.deepin.dde.ControlCenter /com/deepin/dde/ControlCenter com.deepin.dde.ControlCenter.Show");

        // call exist service
//        DBusControlCenter c;

//        if (parser.isSet(toggleOption)) {
//            c.Toggle();
//        }

//        if (!positionalArgs.isEmpty()) {
//            c.ShowModule(positionalArgs.at(0));
//        } else if (parser.isSet(showOption)) {
//            c.Show();
//        }

#ifndef QT_DEBUG
        return 0;
#endif
    }

#ifndef QT_DEBUG
    if (parser.isSet(showOption) || parser.isSet(toggleOption))
#endif
        QTimer::singleShot(1, &f, &Frame::startup);

    if (!positionalArgs.isEmpty())
        QTimer::singleShot(10, &f, [&] { f.showSettingsPage(positionalArgs.first(), QString()); });
//        f.showSettingsPage(positionalArgs.first(), QString());

    return app.exec();
}

//#include <gtk/gtk.h>

//#include <QApplication>
//#include <QFile>
//#include <QDebug>
//#include <QTranslator>
//#include <QCommandLineParser>

//#include "frame.h"
//#include "interfaces.h"
//#include "pluginsmanager.h"
//#include "dbus/dbuscontrolcenter.h"
//#include "dbus/dbuscontrolcenterservice.h"

//#include <dapplication.h>
//#include <dthememanager.h>

//#include <DLog>

//DUTIL_USE_NAMESPACE
//DWIDGET_USE_NAMESPACE

//QString getQssFromFile(const QString &name)
//{
//    QString qss = "";

//    QFile f(name);
//    if (f.open(QIODevice::ReadOnly | QIODevice::Text)) {
//        qss = f.readAll();
//        f.close();
//    }

//    return qss;
//}

//void onThemeChange(const QString &theme)
//{
//    QString fileName = QString(":/resources/qss/%1.qss").arg(theme);

//    qApp->setStyleSheet(getQssFromFile(fileName));
//}

//int main(int argv, char *args[])
//{
//    gtk_init(&argv, &args);

//    DApplication app(argv, args);
//    app.setOrganizationName("deepin");
//    app.setApplicationName("dde-control-center");
//    app.setApplicationVersion("3.0");

//    QFont ft12 = app.font();
//    ft12.setPixelSize(12);
//    app.setFont(ft12);

//    // keep running if all window hided
//    qApp->setQuitOnLastWindowClosed(false);

//#ifndef QT_DEBUG
//    QDir::setCurrent(QApplication::applicationDirPath());
//#endif

//    // initialize logging
//    DLogManager::registerConsoleAppender();
//#ifdef QT_DEBUG
//    DLogManager::registerConsoleAppender();
//#else
//    DLogManager::registerFileAppender();
//#endif

//    qDebug() << "dde-control-center start";

//    // install translators
//    QTranslator translator;
//    translator.load("../share/dde-control-center/translations/dde-control-center_" + QLocale::system().name());
//    app.installTranslator(&translator);

//    // take care of command line options
//    QCommandLineOption showOption(QStringList() << "s" << "show", "show control center(hide for default).");
//    QCommandLineOption toggleOption(QStringList() << "t" << "toggle", "toggle control center visible.");
//    QCommandLineParser parser;
//    parser.setApplicationDescription("DDE Control Center");
//    parser.addHelpOption();
//    parser.addVersionOption();
//    parser.addOption(showOption);
//    parser.addOption(toggleOption);
//    parser.addPositionalArgument("module", "the module's id of which to be shown.");
//    parser.process(app);

//    QStringList positionalArgs = parser.positionalArguments();

//    Frame frame;
//    DBusControlCenterService adaptor(&frame);
//    QDBusConnection conn = QDBusConnection::sessionBus();
//    if (!conn.registerService("com.deepin.dde.ControlCenter") ||
//            !conn.registerObject("/com/deepin/dde/ControlCenter", &frame)) {
//        qDebug() << "dbus service already registered!";

//        // call exist service
//        DBusControlCenter c;

//        if (parser.isSet(toggleOption)) {
//            c.Toggle();
//        }

//        if (!positionalArgs.isEmpty()) {
//            c.ShowModule(positionalArgs.at(0));
//        } else if (parser.isSet(showOption)) {
//            c.Show();
//        }

//#ifndef QT_DEBUG
//        return 0;
//#endif
//    }

//    frame.loadContens();

//    if (!positionalArgs.isEmpty()) {

//        PluginsManager *pluginsManager = PluginsManager::getInstance(&app);

//        if (pluginsManager->pluginIndex(positionalArgs.at(0)) != -1) {
//            frame.selectModule(positionalArgs.at(0));
//        } else {
//            frame.selectModule("home");
//        }
//    }
//#ifndef QT_DEBUG
//    else if (parser.isSet(showOption) || parser.isSet(toggleOption))
//#endif

//        frame.show();
//    // setup theme manager
//    DThemeManager *manager = DThemeManager::instance();
//    QObject::connect(manager, &DThemeManager::themeChanged, [ = ](QString theme) {
//        onThemeChange(theme);
//    });

//    manager->setTheme("dark");
//    onThemeChange("dark");

//    return app.exec();
//}
